libc/benchmarks/gpu/timing/amdgpu/timing.h

```diff
@@ -0,0 +1,73 @@
+//===------------- AMDGPU implementation of timing utils --------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_UTILS_GPU_TIMING_AMDGPU
+#define LLVM_LIBC_UTILS_GPU_TIMING_AMDGPU
+
+#include "src/__support/GPU/utils.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/attributes.h"
+#include "src/__support/macros/config.h"
+
+#include <stdint.h>
+
+namespace LIBC_NAMESPACE {
+
+// Returns the overhead associated with calling the profiling region. This
+// allows us to substract the constant-time overhead from the latency to
+// obtain a true result. This can vary with system load.
+[[gnu::noinline]] static LIBC_INLINE uint64_t overhead() {
+  gpu::memory_fence();
+  uint64_t start = gpu::processor_clock();
+  uint32_t result = 0.0;
+  asm("v_or_b32 %[v_reg], 0, %[v_reg]\n" ::[v_reg] "v"(result) :);
+  asm("" ::"s"(start));
+  uint64_t stop = gpu::processor_clock();
+  return stop - start;
+}
+
+// Profile a simple function and obtain its latency in clock cycles on the
+// system. This function cannot be inlined or else it will disturb the very
+// delicate balance of hard-coded dependencies.
+template <typename F, typename T>
+[[gnu::noinline]] static LIBC_INLINE uint64_t latency(F f, T t) {
+  // We need to store the input somewhere to guarantee that the compiler will
+  // not constant propagate it and remove the profiling region.
+  volatile uint32_t storage = t;
+  float arg = storage;
```

Comment:
This should be `T`, not `float`.

Reply:
Also we need versions with more than one input, right.

Reply:
How many do you think we need? I think I remember you mentioned 6 inputs at max - is that still correct?

Reply:
Just a guess, for now at least two, we can add more as needed. Maybe I just don't know enough C++ magic to automate it since the inputs can be potentially different types. (Plus all that C++ magic needs to exist in the `libc` project)

Reply:
Are there problems with trying to link when it's generic? I didn't notice before but trying to use `T` causes `lld` to throw errors such as `error: couldn't allocate input reg for constraint 'r'`(And for the `s` asm input constraint) when I switched from `uint64_t` to `T`

Reply:
Yeah, the assembly constraints are different for AMDGCN. I believe LLVM uses the same ones as GCC, and their documentation is a bit more complete: https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html.

Reply:
For reference, AMDGCN has separate types of registers while PTX just lists them all as the same. Generally, there's `sgpr`, `vgpr` mainly. There's `agpr` but those are special registers mostly used for the high end cards so I wouldn't worry about them. SGPRs are scalar, they hold values such that they are uniform throughout the wavefront. VGPRs hold values for each lane in the wavefront. If you're familiar with SIMD in your x64 CPU, it's the same concept.

Reply:
Is there a specific constraint that generics need to use, I tried a couple like `v` and `Sg` but `lld` still threw errors - am I missing a step that we need to work with generics since the size isn't known for sure?

Reply:
Never use the r constraint. We should just error on it. 

GCC went and added a bunch of constraints on their own. We don't support all of the ones there.

What do you mean by generics? The constraints don't care what the type is (other than maybe for the weird immediate cases), they just need to be scalar or vector. The SIMD analog doesn't actually work, these shouldn't be thought of as SIMD vectors vs. scalars

Reply:
I guess the short answer is 99% of the time you want `"v"`. There's definitely a lot of weirdness so it's not exactly, VGPRs are more like a big block of resource that can be assigned. I think you can even unassign a VGPR during execution? I should probably find a concise way to explain the magic of SIMT and hardware.

Reply:
> What do you mean by generics? The constraints don't care what the type is (other than maybe for the weird immediate cases), they just need to be scalar or vector. The SIMD analog doesn't actually work, these shouldn't be thought of as SIMD vectors vs. scalars

Ah, I was mixing this up with something else. I think I meant to ask if there are certain conditions that need to be met for these assembly constraints when using templates. `ld` throws an error that it can't allocate the input register for the `v` constraint for my templated local. Is it because the compiler doesn't know how much space the variable could take up?

Reply:
It needs to be a legal, handled type which is about it